src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -80,19 +92,119 @@ public void completed(Integer result, Void v) {
                         }
 
                         @Override
-                        public void failed(Throwable exc, Void v) {
+                        public void failed(Throwable exc, Void ignore) {
                             VolleyLog.e(exc, "Failed to read file %s", file.getAbsolutePath());
+                            closeChannel(afc, "failed read");
                             callback.onGetComplete(null);
                         }
                     });
         } catch (IOException e) {
             VolleyLog.e(e, "Failed to read file %s", file.getAbsolutePath());
+            closeChannel(channel, "IOException");
             callback.onGetComplete(null);
         }
     }
 
-    /** Returns a file object for the given cache key. */
-    File getFileForKey(String key) {
-        return new File(mRootDirectorySupplier.get(), DiskBasedCacheUtility.getFilenameForKey(key));
+    /** Puts the cache entry with a specified key into the cache. */
+    @Override
+    public void put(final String key, Cache.Entry entry, final OnPutCompleteCallback callback) {
+        if (DiskBasedCacheUtility.wouldBePruned(
+                mTotalSize, entry.data.length, mMaxCacheSizeInBytes)) {
+            return;
+        }
+
+        final File file = DiskBasedCacheUtility.getFileForKey(key, mRootDirectorySupplier);
+        Path path = Paths.get(file.getPath());
+
+        // channel we can close after IOException
+        AsynchronousFileChannel channel = null;
+        try {
+            final AsynchronousFileChannel afc =
+                    AsynchronousFileChannel.open(
+                            path, StandardOpenOption.WRITE, StandardOpenOption.CREATE);
+            channel = afc;
+            final CacheHeader header = new CacheHeader(key, entry);
+            int headerSize = header.getHeaderSize();
+            final int size = entry.data.length + headerSize;
+            ByteBuffer buffer = ByteBuffer.allocate(size);
+            header.writeHeader(buffer);
+            buffer.put(entry.data);
+            buffer.flip();
+            afc.write(
+                    /* source= */ buffer,
+                    /* position= */ 0,
+                    /* attachment= */ null,
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            if (closeChannel(afc, "completed read")) {
+                                if (resultLen != size) {
+                                    VolleyLog.e(
+                                            "File changed while reading: %s",
```

Comment:
this is a write, correct?

Also, I think we should delete the file in this case, since it wouldn't have valid contents.